File: Leetcode-Submissions/0367-valid-perfect-square/solution.cpp
bool sqrt(int x){
    int s = 1;
    int e = x;

    long long int mid = s + (e - s)/2;

    while(s <= e){

        long long int sqr = mid * mid;
        if(sqr == x){

         return 1;

    }

       else if ( sqr < x){

           s = mid + 1;

       }

        else{

            e = mid - 1;

        }

        mid = s + (e - s)/2;

    }

    return 0;

    

}

class Solution {
public:
    bool isPerfectSquare(int num) {
        return sqrt(num);
    }
};

File: Leetcode-Submissions/0367-valid-perfect-square/solution.java
class Solution {
    public boolean isPerfectSquare(int num) {
        int l = 1, r = num;
        while(l <= r){
            int mid = l + (r - l) / 2;
            if(mid * mid == num) return true;
            else if(mid < num / mid) l = mid + 1;
            else r = mid - 1;
        }
        return false;
    }
}
